### ISSUE TITLE
High CPU usage/tab crash when sending top-level comment with links

### ISSUE THREAD

**Issue by dato** [Reviewable/Reviewable#758]
Hi—

I don’t know how to best help in diagnosing this, but I was trying to send a review with 19 comments + one top-level comment, and it crashed Chromium and Chrome every time (or, better put, Chrome killed the tab after high CPU usage; I say this because in Firefox, it warned that it was "slowing down the browser", and asked whether to kill it, or wait. I waited a lot, and it just kept on using more RAM).

I deleted the top-level draft, and then hit publish. All code comments went out without a glitch (as always! this is a first). Then I tried to send the top-level comment by itself, and the same still happened (!).

I sent the comment via the Github interface, and then I attempted to narrow down what was triggering the issue on Reviewable. The text of the comment had several links, some to Reviewable itself, and I started by deleting those. Once I did, the comment went out without a glitch (!).

I haven't deleted the duplicate comment in the PR, in case it's useful for you to be able to diagnose. Also, I could paste the links that triggered the issue.

The PR is `fiubatps/algo2_2020a_g07_baffico_oropeza#1` (is this useful? I'm not sure if I should do something more). The two instances of the top-level comments are there, at the end.

Many thanks in advance for any light you might bring to the issue.

**Comment by pkaminski** [comment]
Wow, that's a neat bug you found!  I'm not able to access the private PR, but really all I should need is the URL to make a repro.  Could you paste that here?  Perhaps it's an accidental regex DoS, since otherwise Reviewable doesn't really care about the links in a comment...

**Comment by dato** [comment]
Thanks for taking a look! These are all the URLs, with a bit of context in case it's a tricky one:

```
Bla bla:

  - item one
  - item two linked to [en Github.com](https://github.com/fiubatps/algo2_2020a_g07_baffico_oropeza/pull/1)
  - and item three linked both to [Reviewable.io](https://reviewable.io/) (and then a second link to [reviewable.io/algo2_2020a_g07_baffico_oropeza#1](https://reviewable.io/reviews/fiubatps/algo2_2020a_g07_baffico_oropeza/1#-))

<some more paragraphs>

Last paragraph had a bare link to https://reviewable.io/reviews as naked URL.
```

**Comment by pkaminski** [comment]
Hmm, I tried putting this in a comment and it got sent without problems.  Are you still able to reproduce the issue yourself?  If so, can you check whether the preview for the comment gets generated correctly or not?  And if so, try publishing again (preferably from a clean page load) and look at the contents of `Raven._breadcrumbs` afterwards in the console -- it may give us a rough idea of where the process gets stuck.  Thanks!

**Comment by dato** [comment]
It still happens, and I’m not sure how to help.

First, here's the exact Markdown that I'm pasting on the top-level comment box (I set expiry date to a week): http://dpaste.com/25STPXG.txt

Second, the preview renders without problem (both in the comment, and in the publish drop-down menu.)

Third, I'm not sure how to check `Raven._breakcumbs`. After the crash I mean. I don't think it's possible? (But I'm not very familiar with developer tools, perhaps there's a watch feature I don't know about.) In any case:

  - before hitting publish, I typed `Raven._breadcumbs` in the console, I could see a list that had 6 items (if it's relevant what, I could get them into a file and post the here)

  - during the "sending process" (when the tab hangs), I could type into the console, but well, it didn't reply back :upside_down_face: 

  - after Chromium  kills the tab, the console is disconnected.

-----

Update: I tried with Firefox, which doesn't kill the tab. But well, while it is sending, it gives an empty responseto `Raven._breadcumbs`. But I noticed this in the log:

```
[2020-07-08T02:33:44.401Z]  @firebase/database: FIREBASE WARNING: update at /reviews/-MB6NogJRDR_5yQqkwGV/tracker/-top/participants/github:73768 failed: permission_denied
```

(I have no idea whether it's relevant or not.)

**Comment by dato** [comment]
Ah, wait; further update: After I told Firefox to stop, I got a backtrace, and Raven now had 13 items.

Backtrace:

```
Error: Script terminated by timeout at:
e@https://reviewable.io/js/dab49e8c.libs.js:20:6659
_makeRequest@https://reviewable.io/js/dab49e8c.libs.js:19:25404
_sendProcessedPayload@https://reviewable.io/js/dab49e8c.libs.js:19:24815
_send@https://reviewable.io/js/dab49e8c.libs.js:19:23817
_processException@https://reviewable.io/js/dab49e8c.libs.js:19:20954
_handleStackInfo@https://reviewable.io/js/dab49e8c.libs.js:19:19471
captureException@https://reviewable.io/js/dab49e8c.libs.js:19:8586
@https://reviewable.io/js/9b830254.app.js:13:5665
f.prototype.capture@https://reviewable.io/js/9b830254.app.js:25:18407
g.prototype.publish/<@https://reviewable.io/js/9b830254.app.js:25:6260
h@https://reviewable.io/js/dab49e8c.libs.js:5:6775
i/<@https://reviewable.io/js/dab49e8c.libs.js:5:6947
$eval@https://reviewable.io/js/dab49e8c.libs.js:5:14430
$digest@https://reviewable.io/js/dab49e8c.libs.js:5:12903
@https://reviewable.io/js/dab49e8c.libs.js:14:8430
ra/<@https://reviewable.io/js/dab49e8c.libs.js:11:30721
qa@https://reviewable.io/js/dab49e8c.libs.js:11:30662
raven.js:3815:13
```

Breadcumbs (not sure if there's a better way to download/inspect this data):

```
(13) […]
​
0: {…}
​​
category: "action"
​​
data: Object { title: "Pull Request Review", windowWidth: 1050, windowHeight: 837, … }
​​
message: "Viewed Review Page"
​​
timestamp: 1594175593.882
​​
<prototype>: Object { … }
​
1: Object { timestamp: 1594175594.217, message: "Download the Vue Devtools extension for a better development experience:\nhttps://github.com/vuejs/vue-devtools", level: "info", … }
​
2: {…}
​​
category: "connection"
​​
message: "Disconnected"
​​
timestamp: 1594175594.282
​​
<prototype>: Object { … }
​
3: {…}
​​
category: "auth"
​​
data: Object { uid: "github:73768", emailVerified: false, isAnonymous: false, … }
​​
timestamp: 1594175595.976
​​
<prototype>: Object { … }
​
4: {…}
​​
category: "warden"
​​
message: "Pinged permission ticket /queues/permissions/fiubatps|algo2_2020a_g07_baffico_oropeza|github:73768"
​​
timestamp: 1594175595.994
​​
<prototype>: Object { … }
​
5: {…}
​​
category: "connection"
​​
message: "Connected"
​​
timestamp: 1594175596.589
​​
<prototype>: Object { … }
​
6: {…}
​​
category: "navigation"
​​
data: Object { to: "/reviews/fiubatps/algo2_2020a_g07_baffico_oropeza/1#-", from: "/reviews/fiubatps/algo2_2020a_g07_baffico_oropeza/1" }
​​
timestamp: 1594175606.308
​​
<prototype>: Object { … }
​
7: {…}
​​
category: "ui.click"
​​
message: "div.ui.bottom.tabular.draft.menu > a.preview.item"
​​
timestamp: 1594175606.36
​​
<prototype>: Object { … }
​
8: {…}
​​
category: "ui.click"
​​
message: "div.ui.bottom.tabular.draft.menu > a.write.item"
​​
timestamp: 1594175607.3
​​
<prototype>: Object { … }
​
9: {…}
​​
category: "ui.click"
​​
message: "span.ui.dimmable.green.publish.buttons > div.ui.button"
​​
timestamp: 1594175623.011
​​
<prototype>: Object { … }
​
10: {…}
​​
category: "process"
​​
data: Object { numDrafts: 1, numUnpostedDrafts: 1, postToGitHub: true }
​​
message: "Publishing comments"
​​
timestamp: 1594175623.52
​​
<prototype>: Object { … }
​
11: {…}
​​
category: "process"
​​
data: Object { directives: {…}, requestCurrentUser: false }
​​
message: "Analyzed directives"
​​
timestamp: 1594175623.99
​​
<prototype>: Object { … }
​
12: {…}
​​
category: "sentry"
​​
event_id: "fa32fc380028460388a72b7e906b5f7e"
​​
level: "warning"
​​
message: "Error: PERMISSION_DENIED: Permission denied"
​​
timestamp: 1594175624.428
​​
<prototype>: Object { … }
​
length: 13
​
<prototype>: Array []

```

**Comment by dato** [comment]
One more datapoint: I tried posting the same comment in another PR under a different repository in the same organization: still happens.

Just let me know if there's any kind of test, or access, that could be useful to you. I ran out of things to try. =)

**Comment by pkaminski** [comment]
I wasn't able to reproduce with the given message in a couple different repos in `reviewable`, so the problem isn't just the content.  Since you say it's not repo-specific either, it's likely that it's specific to the organization.  I wonder if it's some variant of #755; can you see if any network requests are emitted while Reviewable is stuck?  If so, I have a fix nearly ready to go that may help.  If possible, could you also try posting in a different organization (e.g., in  [reviewable/demo](https://reviewable.io/reviews/Reviewable/demo/1#-) to make sure that this isn't specific to you somehow?

Thanks for capturing the breadcrumbs.  They indicate that the issue is happening pretty late in the publishing sequence.  The permission failure is confusing: it should be PR-specific, so if you see it on different PRs then my mental model is off somehow.  (It's most likely being caused by the self-mention somehow.)

The stack trace appears to point to the exception capture routine but it's quite unlikely this would get stuck, so it's probably just chance that it happened to break the execution there -- there's lots of other stuff going on in parallel promises.

For now I'll focus on getting the fix to #755 ready to ship tomorrow, as that will at least eliminate one potential trouble spot.

**Comment by dato** [comment]
This gets weirder by the minute. I was able to post to reviewable/demo successfully, then I realized I had done it from my Personal Chrome profile (not my Uni) profile. Went to my Uni profile to post it from there, it hanged. Want back to my Personal profile, it hanged as well. (But, alas, the draft from the Uni profile was there, so it could have interfered.)

I'll see about testing this in more detail tomorrow, in some other org.

**Comment by dato** [comment]
(To be clear, I was able to post to demo#1 exactly once.)

**Comment by pkaminski** [comment]
I released the fix for #755, could you please check if that affected your repro at all?  Thanks.

**Comment by dato** [comment]
I was not able to post the text of http://dpaste.com/25STPXG.txt nor in (a clean load of) the original PR, nor in Reviewable/demo#1 (I had a draft there, so I posted  dummy comment to make sure I could have a real clean re-load). Chromium terminated both tabs (in different profiles fwiw).

**Comment by pkaminski** [comment]
This is really quite weird.  Sorry to keep bugging you but since you're the only one who can repro so far:  does it happen if you remove the @dato self-mention from the message?

**Comment by dato** [comment]
>Sorry to keep bugging you

On the contrary, happy to help and that you're happy to keep looking into the issue!

>does it happen if you remove the @dato self-mention from the message?

No, it doesn't! I was able to post the same text, without the last sentence, to both Reviewable/demo#1, and the original pull request. (Then I attempted it again, with the full text, to confirm that it still hanged, and it does—which is good, I guess.)

**Comment by pkaminski** [comment]
So for some reason I still can't reproduce the issue, but all the hints were sufficient to attempt a 2-line fix.  It's deployed now -- could you give it another try?  Thanks!

**Comment by dato** [comment]
I was able to post the comment twice on each the original pull request, and Reviewable/demo#1! :tada: 

Thank you for your commitment to investigating and solving it. :slightly_smiling_face: 

(Out of curiosity: what was the fixed issue?)

**Comment by pkaminski** [comment]
The root cause was that self-mentions were not allowed by Reviewable's internal schema, but the client code blindly processed them anyway.  The mystery, though, is that this should've caused an error to be raised and propagated all the way up to the UI rather than just spinning.  Still not sure what went wrong there...
